Support [LegacyWindowAlias]

saschanaz opened the pull request:
<!-- Please describe your changes on the following line: -->

---
<!-- Thank you for contributing to Servo! Please replace each `[ ]` by `[X]` when the step is complete, and replace `___` with appropriate data: -->
- [x] `./mach build -d` does not report any errors
- [x] `./mach test-tidy` does not report any errors
- [x] These changes fix #24474 (GitHub issue number if applicable)

<!-- Either: -->
- [x] There are tests for these changes

<!-- Also, please make sure that "Allow edits from maintainers" checkbox is checked, so that we can help you if you get stuck somewhere along the way.-->

<!-- Pull requests that do not address these steps are welcome, but they will require additional verification as part of the review process. -->

saschanaz committed:
Support [LegacyWindowAlias]

highfive commented:
Heads up! This PR modifies the following files:
 * @asajeffrey: components/script/dom/bindings/codegen/CodegenRust.py, components/script/dom/webidls/URL.webidl, components/script/dom/bindings/interface.rs
 * @kichjang: components/script/dom/bindings/codegen/CodegenRust.py, components/script/dom/webidls/URL.webidl, components/script/dom/bindings/interface.rs

saschanaz commented:
@bors-servo try=wpt

bors-servo commented:
:hourglass: Trying commit 27c47b461c7ed609ad9aae181cf25e5f22ae8387 with merge fde26f4519b24fd351ad4942fe8d0fcee0ad1d2f...

bors-servo commented:
:broken_heart: Test failed - [linux-rel-wpt](http://build.servo.org/builders/linux-rel-wpt/builds/13824)

saschanaz commented:
@bors-servo try=wpt

bors-servo commented:
:hourglass: Trying commit f47089267b69f9be06278cfab2a3d678645cd3b6 with merge 500c4acd97af4330e6859eb3ea9c049a6dd7222e...

bors-servo commented:
:sunny: Test successful - [linux-rel-css](http://build.servo.org/builders/linux-rel-css/builds/13763), [linux-rel-wpt](http://build.servo.org/builders/linux-rel-wpt/builds/13827), [status-taskcluster](https://tools.taskcluster.net/task-group-inspector/#/PTDhA9SXSlyjLGH-CE1ePg)
State: approved= try=True

nox commented on components/script/dom/bindings/interface.rs:
```diff
@@ -260,6 +261,12 @@ pub fn create_noncallback_interface_object(
     define_length(cx, rval.handle(), i32::try_from(length).expect("overflow"));
     define_on_global_object(cx, global, name, rval.handle());
+
+    if is_exposed_in(global, Globals::WINDOW) {
```
This attribute is forbidden in non-Window globals so I'm not sure what this test is for.

saschanaz replied:
Interfaces with `[Exposed=(Window, Worker)]` can also have legacy window aliases, while they only actually get aliases on Window. This is to match the specified behavior.

nox reviewed (commented):
Just one question, r=me otherwise. Thanks for your contribution!

nox commented:
@bors-servo r+

bors-servo commented:
:pushpin: Commit f470892 has been approved by `nox`

<!-- @bors-servo r=nox f47089267b69f9be06278cfab2a3d678645cd3b6 -->

bors-servo commented:
:hourglass: Testing commit f47089267b69f9be06278cfab2a3d678645cd3b6 with merge 873b1a661334a9d1401a3a0ad5a1660da8dce872...

saschanaz commented:
Sorry to push again, fixed a test-tidy error, interface map data generation (this caused a bug in WebKitCSSMatrix test), and also added WebKitCSSMatrix.

nox commented:
No worries.

@bors-servo r+

bors-servo commented:
:pushpin: Commit e78c3a6 has been approved by `nox`

<!-- @bors-servo r=nox e78c3a6a902082fa4bd41e0bea9d1f087d013b80 -->

bors-servo commented:
:hourglass: Testing commit e78c3a6a902082fa4bd41e0bea9d1f087d013b80 with merge f49f349df55e8c9798a4c06adc949efba1f81f15...

jdm commented:
This needs `mach update-manifest` now.

jdm commented:
@bors-servo r+

bors-servo commented:
:pushpin: Commit fc96f30 has been approved by `jdm`

<!-- @bors-servo r=jdm fc96f30c1d0cd324cb040e682d689f6a0d808a1e -->

bors-servo commented:
:hourglass: Testing commit fc96f30c1d0cd324cb040e682d689f6a0d808a1e with merge 5b4d96df47872095c5b054783fa8e875bac05d95...

bors-servo commented:
:broken_heart: Test failed - [linux-rel-css](http://build.servo.org/builders/linux-rel-css/builds/13801)

bors-servo commented:
:umbrella: The latest upstream changes (presumably #24476) made this pull request unmergeable. Please resolve the merge conflicts.

jdm commented on tests/wpt/mozilla/tests/mozilla/interfaces.js:
```diff
@@ -45,9 +45,9 @@ function test_interfaces(interfaceNamesInGlobalScope) {
-      "Uint8ClampedArray",
       "WeakMap",
       "WeakSet",
+      "WebKitCSSMatrix",
```
Whoops; this belongs in interfaces.html instead, and with a lowercase K.

saschanaz replied:
Whoooops!! Fixed it. It's `WebKitCSSMatrix` anyway though, as WebKit people somehow decided to use uppercase K here.

jdm commented:
@bors-servo r+

bors-servo commented:
:pushpin: Commit e81b678 has been approved by `jdm`

<!-- @bors-servo r=jdm e81b6786457b076a06c945835a3080fa33fa81ca -->

bors-servo commented:
:hourglass: Testing commit e81b6786457b076a06c945835a3080fa33fa81ca with merge 605ddbecd4dfbbb67849cf63774b2d7635fc3601...

bors-servo commented:
:sunny: Test successful - [linux-rel-css](http://build.servo.org/builders/linux-rel-css/builds/13808), [linux-rel-wpt](http://build.servo.org/builders/linux-rel-wpt/builds/13872), [status-taskcluster](https://tools.taskcluster.net/task-group-inspector/#/cgULlYjcRIC47Y-Izce0ew)
Approved by: jdm
Pushing 605ddbecd4dfbbb67849cf63774b2d7635fc3601 to master...
